raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -360,26 +359,30 @@ private void maybeBecomeLeader(CandidateState state) throws IOException {
         }
     }
 
-    private void onBecomeCandidate(CandidateState state) throws IOException {
+    private void onBecomeCandidate(CandidateState state, int retries) throws IOException {
         maybeBecomeLeader(state);
         resetConnections();
 
         kafkaRaftMetrics.updateElectionStartMs(time.milliseconds());
-        timer.reset(electionTimeoutMs + randomElectionJitterMs());
+        timer.reset(randomElectionTimeoutMs());
     }
 
-    private void becomeCandidate() throws IOException {
+    private void becomeCandidate(int retries) throws IOException {
         // after we become candidate, we may not fetch from leader any more
-        CandidateState state = quorum.becomeCandidate();
-        onBecomeCandidate(state);
+        CandidateState state = quorum.becomeCandidate(retries);
+        onBecomeCandidate(state, retries);
     }
 
     private void becomeUnattachedFollower(int epoch) throws IOException {
         if (quorum.becomeUnattachedFollower(epoch)) {
+            // if we are voter, becoming unattached also means that we would no longer fetch
+            // from whoever the old leader already, while there's no new leader learned yet.
+            // So this member should qualify to elect as leader as well, and hence we shorten
+            // the timer to election timeout;
+            //
+            // if we are observer, nothing needed to be done as we would still try to find-quorum eventually
```

Comment:
It would be helpful to have some documentation somewhere explaining the timeout logic at a higher level. My basic understanding is that we have the base election timeout, which is set randomly in the interval (T, 2T) where T is controlled by `election.timeout.ms`. Then we have the backoff behavior, which begins at `RETRY_BACKOFF_BASE_MS` and increases exponentially up to `election.jitter.ms`. Is that right?

We don't have to do it here, but I think the timeout logic is complex enough that we might want to move it into a separate class for easier testing. For example, we could have a separate `ElectionTimer` which encapsulates the `Timer` instance and provides higher level APIs. For example:

```java
class ElectionTimer {
 private final Timer timer;

 void onBecomeCandidate();
 void onBecomeLeader();
 void onElectionFailed();
...
```

Reply:
Yes the understanding is correct, I will add some more docs for it.

Regarding to move it to a separate class, I'm a bit torn on this, since we've just spent some time consolidating the timeout mechanism into a single place, and the main idea in this PR is still to use that timer for all 1) leader fetch timeout, 2) election timeout, 3) election backoff, 4) voter fetch timeout.

Reply:
Yeah, I'm not necessarily suggesting that we introduce multiple timers. We can keep one consolidated timer. The main thing I'm suggesting is pulling the logic into a separate class to make it easier to test. Anyway, we can consider this as a follow-up if you like.

Reply:
Yeah, I wasn't necessarily suggesting that we separate the timers. We can keep them consolidated, but still separate the logic to reset them in a separate class. Anyway, I don't feel too strongly about it. We can consider it for a follow-up if you want.

Reply:
I thought about it and refactored the code a bit to extract the time generated for each `timer.reset` call into a separate function and added some docs as well. LMK if you think that's clear enough for readers.